File: rlox/lox/scanner.rb
require_relative "./token"

class Scanner
  def initialize()
    starting_state(nil)
  end

  def scan_tokens(source)
    starting_state(source)

    while !is_at_end()
      @start_pos = @current_pos
      scan_token()
    end

    @start_pos = @current_pos
    add_token(:EOF)

    @tokens
  end

  private #=====================================================================

  def is_at_end()
    @current_pos >= @source.length
  end

  def scan_token()
    char = advance()
    case char
    when "("; add_token(:LEFT_PAREN)
    when ")"; add_token(:RIGHT_PAREN)
    when "{"; add_token(:LEFT_BRACE)
    when "}"; add_token(:RIGHT_BRACE)
    when ","; add_token(:COMMA)
    when "."; add_token(:DOT)
    when "-"; add_token(:MINUS)
    when "+"; add_token(:PLUS)
    when ";"; add_token(:SEMICOLON)
    when "*"; add_token(:STAR)
    when "!"; match("=") ? add_token(:BANG_EQUAL) : add_token(:BANG)
    when "="; match("=") ? add_token(:EQUAL_EQUAL) : add_token(:EQUAL)
    when "<"; match("=") ? add_token(:LESS_EQUAL) : add_token(:LESS)
    when ">"; match("=") ? add_token(:GREATER_EQUAL) : add_token(:GREATER)
    when "/"; match("/") ? comment() : add_token(:SLASH)
    when pred(:whitespace?); @tokens
    when "\n"; @line_num += 1; @tokens
    when "\""; string()
    when pred(:digit?); number()
    when pred(:alpha?); identifier()
    else Lox.error(@line_num, "Unexpected character #{char}.")
    end
  end

  def whitespace?(char)
    [" ", "\r", "\t"].member?(char)
  end

  def digit?(char)
    ("0".."9").member?(char)
  end

  def alpha?(char)
    (char >= "a" && char <= "z") ||
    (char >= "A" && char <= "Z") ||
    char == "_"
  end

  def alpha_numeric?(char)
    alpha?(char) || digit?(char)
  end

  def comment()
    advance() while peek() != "\n" && !is_at_end()
    @tokens
  end

  def string()
    while peek() != "\"" && !is_at_end()
      @line_num += 1 if peek() == "\n"
      advance()
    end

    if is_at_end()
      Lox.error(@line_num, "Unterminated string.")
      return @tokens
    end

    advance() # Consume the closing ".

    add_token(:STRING, token_text()[1...-1])
  end

  def number()
    advance() while digit?(peek())

    # Look for a fractional part.
    if peek() == "." && digit?(peek_next())
      advance() # Consume the dot.
      advance() while digit?(peek())
    end

    add_token(:NUMBER, token_text().to_f)
  end

  def identifier()
    advance() while alpha_numeric?(peek())

    keywords = %w(
      and class else false for fun if nil or
      print return super this true var while
    )
    text = token_text()
    type = keywords.member?(text) ? text.upcase.to_sym : :IDENTIFIER

    add_token(type)
  end

  def token_text()
    @source[@start_pos...@current_pos]
  end

  def add_token(type, literal = nil)
    @tokens.push(Token.new(type, token_text(), literal, @line_num))
  end

  def advance()
    char = @source[@current_pos] || ""
    @current_pos += 1
    char
  end

  def match(expected)
    return false if is_at_end()
    return false if @source[@current_pos] != expected

    @current_pos += 1
    true
  end

  def peek()
    @source[@current_pos] || ""
  end

  def peek_next()
    @source[@current_pos + 1] || ""
  end

  def pred(method_name)
    lambda(&method(method_name))
  end

  def starting_state(source)
    @source = source
    @tokens = []
    @start_pos = 0
    @current_pos = 0
    @line_num = 1
  end
end

File: zlox/src/scanner.zig
const std = @import("std");
const Token = @import("./token.zig").Token;
const TokenType = @import("./token.zig").TokenType;

pub const Scanner = struct {
    source: []const u8,
    current: usize,
    line: u32,

    pub fn init(source: []const u8) Scanner {
        return Scanner{
            .source = source,
            .current = 0,
            .line = 1,
        };
    }

    pub fn scanToken(self: *Scanner) Token {
        self.skipWhitespace();

        self.source = self.source[self.current..];
        self.current = 0;

        if (self.isAtEnd()) return self.makeToken(.Eof);

        const c = self.advance();

        return switch (c) {
            '(' => self.makeToken(.LeftParen),
            ')' => self.makeToken(.RightParen),
            '{' => self.makeToken(.LeftBrace),
            '}' => self.makeToken(.RightBrace),
            ';' => self.makeToken(.Semicolon),
            ',' => self.makeToken(.Comma),
            '.' => self.makeToken(.Dot),
            '-' => self.makeToken(.Minus),
            '+' => self.makeToken(.Plus),
            '/' => self.makeToken(.Slash),
            '*' => self.makeToken(.Star),
            '!' => self.makeToken(if (self.match('=')) .BangEqual else .Bang),
            '=' => self.makeToken(if (self.match('=')) .EqualEqual else .Equal),
            '<' => self.makeToken(if (self.match('=')) .LessEqual else .Less),
            '>' => self.makeToken(if (self.match('=')) .GreaterEqual else .Greater),
            '"' => self.string(),
            else => {
                if (isAlpha(c)) return self.identifier();
                if (isDigit(c)) return self.number();
                if (c == 0) return self.makeToken(.Eof);
                return self.errorToken("Unexpected character.");
            },
        };
    }

    fn isAtEnd(self: *Scanner) bool {
        return self.current >= self.source.len;
    }

    fn advance(self: *Scanner) u8 {
        const c = self.peek();
        self.current += 1;
        return c;
    }

    fn skip(self: *Scanner) void {
        self.current += 1;
    }

    fn skipLine(self: *Scanner) void {
        while (self.peek() != '\n' and !self.isAtEnd()) self.skip();
    }

    fn peek(self: *Scanner) u8 {
        if (self.isAtEnd()) return 0;
        return self.source[self.current];
    }

    fn peekNext(self: *Scanner) u8 {
        if (self.isAtEnd()) return 0;
        return self.source[self.current + 1];
    }

    fn match(self: *Scanner, char: u8) bool {
        if (self.isAtEnd() or self.peek() != char) return false;
        self.skip();
        return true;
    }

    fn makeToken(self: *Scanner, tokenType: TokenType) Token {
        return Token{
            .tokenType = tokenType,
            .lexeme = self.source[0..self.current],
            .line = self.line,
        };
    }

    fn errorToken(self: *Scanner, message: []const u8) Token {
        return Token{
            .tokenType = .Error,
            .lexeme = message,
            .line = self.line,
        };
    }

    fn skipWhitespace(self: *Scanner) void {
        while (true) {
            switch (self.peek()) {
                ' ', '\r', '\t' => self.skip(),
                '\n' => {
                    self.line += 1;
                    self.skip();
                },
                '/' => if (self.peekNext() == '/') self.skipLine() else return,
                else => return,
            }
        }
    }

    fn identifier(self: *Scanner) Token {
        while (isAlpha(self.peek()) or isDigit(self.peek())) self.skip();
        return self.makeToken(self.identifierType());
    }

    fn number(self: *Scanner) Token {
        while (isDigit(self.peek())) self.skip();

        if (self.peek() == '.' and isDigit(self.peekNext())) {
            // Consume the ".".
            self.skip();
            while (isDigit(self.peek())) self.skip();
        }

        return self.makeToken(.Number);
    }

    fn string(self: *Scanner) Token {
        while (self.peek() != '"' and !self.isAtEnd()) {
            if (self.peek() == '\n') self.line += 1;
            self.skip();
        }

        if (self.isAtEnd()) return self.errorToken("Unterminated string.");

        // The closing quote
        self.skip();
        return self.makeToken(.String);
    }

    fn identifierType(self: *Scanner) TokenType {
        return switch (self.source[0]) {
            'a' => self.checkKeyword(1, "nd", .And),
            'c' => self.checkKeyword(1, "lass", .Class),
            'e' => self.checkKeyword(1, "lse", .Else),
            'f' => switch (self.source[1]) {
                'a' => self.checkKeyword(2, "lse", .False),
                'o' => self.checkKeyword(2, "r", .For),
                'u' => self.checkKeyword(2, "n", .Fun),
                else => .Identifier,
            },
            'i' => self.checkKeyword(1, "f", .If),
            'n' => self.checkKeyword(1, "il", .Nil),
            'o' => self.checkKeyword(1, "r", .Or),
            'p' => self.checkKeyword(1, "rint", .Print),
            'r' => self.checkKeyword(1, "eturn", .Return),
            's' => self.checkKeyword(1, "uper", .Super),
            't' => switch (self.source[1]) {
                'h' => self.checkKeyword(2, "is", .This),
                'r' => self.checkKeyword(2, "ue", .True),
                else => .Identifier,
            },
            'v' => self.checkKeyword(1, "ar", .Var),
            'w' => self.checkKeyword(1, "hile", .While),
            else => .Identifier,
        };
    }

    fn checkKeyword(self: *Scanner, offset: u8, str: []const u8, tokenType: TokenType) TokenType {
        if (self.current != str.len + offset) return .Identifier;
        const sourceSlice = self.source[offset..self.current];
        std.debug.assert(sourceSlice.len == str.len);
        return if (std.mem.eql(u8, sourceSlice, str)) tokenType else .Identifier;
    }
};

fn isDigit(c: u8) bool {
    return c >= '0' and c <= '9';
}

fn isAlpha(c: u8) bool {
    return (c >= 'a' and c <= 'z') or
        (c >= 'A' and c <= 'Z') or
        c == '_';
}
